### ISSUE TITLE
travis-build: ulimit core addon 

### ISSUE THREAD

**Issue by svenfuchs** [travis-ci/travis-ci#3754]
From support:

> I’d really like to be able to capture C++ backtraces on sudo:false machines, and need some guidance about how best to do this.
> 
> The background is that for C++ applications it is important to be able to get backtrace if the program crashes (to know why it crashed). To get backtraces on linux you need core files. To get core files on linux you need the kernel to have them enabled. For this you need sudo. So I’m realizing today that the method I use at this example repo: https://github.com/springmeyer/travis-coredump only works for `sudo:required` machines. This is because the `ulimit` command appears to not work/ has no effect on `sudo:false` machines. Possible ideas:
> - All travis docker machines default to using `ulimit -c unlimited` when `language: cpp` is used
> - An `addon` is created for `ulimit` that could look like:
> 
> ```
> addons:
>   ulimit:
>     core: unlimited
> ```

I think adding an addon might be the right thing to do. However, is this going to be properly guarded by Docker? Would other containers be somehow affected by this?

/cc @meatballhat @BanzaiMan @joshk

**Comment by BanzaiMan** [comment]
What is the rationale for limiting the core dump size at the moment? If there is no good reason, perhaps setting `unlimited` in general is a better approach than giving an option.

**Comment by springmeyer** [comment]
Thanks so much for giving this consideration. I did a little more debugging last night after sending my first email:

  - I've noticed that the `ulimit` command actually seems to be working: it seems that the process's core file limit has been changed okay because `ulimit -c` reports `unlimited` after running `ulimit -c unlimited -S`.
  - All I really know is that I can't **find** the core file. It may or may not be being created on the system.


The `/proc/sys/kernel/core_pattern` value seems to indicate that apports is set up to intercept core files. So I now suspect that the core files are being swallowed maybe? Perhaps all that needs to be done is to have `apports` removed? Apports does not appear to be running on the docker machines anyway.

For reference I'm experimenting with trying to get core files generated (no success yet) over at https://github.com/springmeyer/travis-coredump/compare/sudo-false and https://travis-ci.org/springmeyer/travis-coredump (in the `sudo-false`) branch. The master branch of that repo that uses `sudo:required` works - the confirmation is that you see the backtrace printed.

More context is that I have experience with enabling core files on ubuntu machines on aws, but I'm not very familiar with docker. Reading https://github.com/docker/docker/issues/11740 to try to understand more.

**Comment by jfirebaugh** [comment]
The ubuntu default for `/proc/sys/kernel/core_pattern` is `|/usr/share/apport/apport %p %s %c %P`, but `/usr/share/apport/apport` is not installed inside the container.

Furthermore, in the unprivileged container environment it is impossible to change `/proc/sys/kernel/core_pattern`. So I don't think there is anything an end user can do to enable core files -- this requires a change to the container by Travis.

**Comment by rphillips** [comment]
We got hit by this as well. Would be nice if ```core_pattern``` is updated to ```core.%e.%p.%h.%t```

**Comment by BanzaiMan** [comment]
I'm not sure if changing `ulimit` would work, but changing the core file location may be feasible.

**Comment by rphillips** [comment]
ulimit can be changed within the ```before_script:``` block. the core file location should be the only fix needed.

**Comment by BanzaiMan** [comment]
On a second thought, creating and making writable the expected core file location seems to be an easy way out.

**Comment by BanzaiMan** [comment]
Maintaining an addon in this case seems like an overkill to me, to be quite honest. If `ulimit -c` is effective (@springmeyer reported—which I overlooked earlier), then it is sufficient to ensure that the jobs can write to it. And set `ulimit -c unlimited` in general.

I think the surprise is that the file isn't written at all, and that, even when the limit is changed, it cannot be written to.

If a user wants to limit the core file size, they can do that at will.

**Comment by joshk** [comment]
So my gut reaction has me against this change as this is what the `sudo: required` setup is for.

On top of this, we have a new setup which is nearly ready, which `ulimit` changes will work on.

I would recommend for now that we close this issue.

**Comment by springmeyer** [comment]
@joshk - I hope before closing this issue we can clarify the problem and agree on ideal steps. The challenge is fundamentally that it is unclear why coredumps are not working on `sudo:false` machines. And only those with the ability to change the containers are able to debug.

As someone who cannot change the containers, my current hunch is the same as @jfirebaugh and @rphillips above: that the problem is that the `core_pattern` pattern default needs to be set for the containers such that apport is not invoked. If that is fixed then the next problem might be `need to be able to set ulimit`. But we cannot know for sure without testing. Can you help?

**Comment by rphillips** [comment]
the ulimit appears to be settable via the user: https://travis-ci.org/luvit/luvit/builds/72494377#L117

**Comment by pnorman** [comment]
> the ulimit appears to be settable via the user: https://travis-ci.org/luvit/luvit/builds/72494377#L117

As discussed above, setting ulimit is not sufficient.

**Comment by cejkato2** [comment]
Has anyone found a way how to change ```core_pattern```?
As @rphillips mentioned, I would need at least %e in order to match core files with executables, something like this:
```
sudo bash -c 'echo %e.core.%p > /proc/sys/kernel/core_pattern'
... make check ...
for i in $(find -name '*.core.*'); do
    ( cd `dirname "$i"`; c=`basename "$i"`
    gdb "${c%%.*}" "$c" -ex "thread apply all bt" -ex "set pagination 0" -batch
    rm "$c";);
done;
```
Unfortunately, I get ```bash: /proc/sys/kernel/core_pattern: Permission denied``` from travis-ci.

**Comment by pnorman** [comment]
> Has anyone found a way how to change core_pattern?

You can't do this on a containerized build.

**Comment by kkm000** [comment]
@cejkato2: `/proc/sys/kernel/core_pattern` is writable only by `root`. The reason is that it can indeed be used to run arbitrary code.

@joshk: 
> So my gut reaction has me against this change as this is what the sudo: required setup is for.
`ulimit` is a user setting (`u` is for user). `ulimit` is a bash builtin, and the value is inherited by all forked processes.

The real problem is that the kernel parameter `kernel.core_pattern` is set to pipe the core dump to a special process (Ubuntu 12 and 14 default is `|/usr/share/apport/apport %p %s %c %P`). This is absolutely not what one would want on a CI worker. apport is for sending crash dumps back to Ubuntu to analyze; this is, generally, a thing for a desktop nonprogrammer user.

Generally, apport would just save the core as a file named `core` but not in a container environment. 

The Linux default for `kernel.core_pattern` is just `core`. Apport is installed, among other files, with `/etc/init.d/apport` that changes the default on start of the service. With upstart, a different file is used, `/etc/init/apport.conf` (I do not remember if Precise comes with init.rc or upstart). My version of Ubuntu has this guard in `/etc/init/apport.conf`:
```
pre-start script
    # don't start in containers
    /bin/running-in-container && exit 0
```

This is what happens on a Ubuntu 14.04 machine when the apport service is started and stopped:
```
root@yupana:# service apport stop
apport stop/waiting
root@yupana:# cat /proc/sys/kernel/core_pattern
core
root@yupana:# service apport start
apport start/running
root@yupana:# cat /proc/sys/kernel/core_pattern
|/usr/share/apport/apport %p %s %c %P
```

Apport can be simply uninstalled with `apt`
```
root@yupana:# cat /proc/sys/kernel/core_pattern
|/usr/share/apport/apport %p %s %c %P
root@yupana:# apt-get remove apport
 . . . churn churn churn . . .
root@yupana:# cat /proc/sys/kernel/core_pattern
core
```

**Comment by cejkato2** [comment]
Thanks @kkm000 for comments, however, I think it is quite strange that ```sudo apt-get install gdb``` works and ```sudo bash -c 'echo %e.core.%p > /proc/sys/kernel/core_pattern'``` doesn't... (I also tried sudo sysctl)
However, according to my experiments, it is actually set to ```core``` (taken from travis' output):
```
$ cat /proc/sys/kernel/core_pattern
core
```

I need gdb(1), which is unfortunately not included by default. Therefore, I use sudo.
As I understand this, usage of sudo (for package installation) prevents me from usage of containers, am I right?
I'm not sure if it is relevant - I found this: ```This job ran on our legacy infrastructure.``` and ```Using worker: worker-linux-3da96de7-2.bb.travis-ci.org:travis-linux-4```

In fact, I've figured it out by a workaround using file(1) and cutting the name of binary (argv[0]) from its output: https://github.com/cejkato2/Nemea-Framework/blob/master/.travis.yml#L44
Anyway, I think this solution is not the one I like because it would be better with changed core_pattern...

**Comment by kkm000** [comment]
@cejkato2:
> However, according to my experiments, it is actually set to core (taken from travis' output):

We are running on the container infra, and it is set to the apport default (see https://travis-ci.org/kaldi-asr/kaldi/builds/76355805#L256). No apport in the container though, so the Travis container configuration is indeed borked :disappointed:.

> I need gdb(1), which is unfortunately not included by default. Therefore, I use sudo.

Can you use the apt addon? See https://github.com/workflow-demo-org/kaldi/blob/9015dce844a1849d2a55a3f5653f736b985ee61e/.travis.yml#L14 

I cannot understand why your attempt to change the parameter fail.

**Comment by cejkato2** [comment]
Interesting, I've tried the apt addon as @kkm000 suggested and now I have core_pattern with apport too.

**Comment by Seadragon91** [comment]
Any updates on this?

**Comment by grondo** [comment]
Can `/proc/sys/kernel/core_pattern` be set to `"core"` in the container host please? That will give users a chance to analyze corefiles generated during their build & test with an `after_failure` hook.

Barring that, I've opened travis-ci/apt-package-whitelist#2642 to have the apport package whitelisted for APT in hopes that, when installed, apport will at least leave a crash report under /var/crash or something (Though there is currently no way to test this).

**Comment by springmeyer** [comment]
I've confirmed that installing `apport` fixes the original problem for me on `sudo:false` machines. Thanks @grondo for that discovery. https://github.com/springmeyer/travis-coredump/blob/master/.travis.yml now offers a working example.

**Comment by springmeyer** [comment]
> I've confirmed that installing apport fixes the original problem for me on sudo:false machines. Thanks @grondo for that discovery. https://github.com/springmeyer/travis-coredump/blob/f57d1ce1e25c66d23dd71030bfc7f009520b27ba/.travis.yml now offers a working example that leaves a core file in the current working directory: https://travis-ci.org/springmeyer/travis-coredump/jobs/144743841#L297

Something has changed and what worked no longer works. Previously a file named `core` was dropped in the current working directory. I just restarted a job that previously worked (and that happened) and now no `core` is generated. So, despite setting `ulimit -c unlimited` and installing `apport` the sudo:false/precise machines now appear broken and don't produce coredumps.

**Comment by grondo** [comment]
I'm seeing the same thing @springmeyer. I think perhaps the `core_pattern` on host machine has changed, and now includes an extra format specifier `%P` at the end of the core pattern:

```
|/usr/share/apport/apport %p %s %c %P
```

Some preliminary testing under 12.04 showed that the version of apport there isn't able to process this extra argument, and instead of ingesting the core file it just bails out with a usage message:

e.g.
```
/usr/share/apport/apport 3290 3 18446744073709551615 2344
Usage: /usr/share/apport/apport <pid> <signal number> <core file ulimit>
The core dump is read from stdin.
```

In my docker image the error is at least logged on the console:
```
[8839305.344433] Core dump to |/usr/share/apport/apport 3290 3 18446744073709551615 2344 pipe failed
```

Under Travis-CI I see no such error, so at this point the above is only conjecture.

**Comment by springmeyer** [comment]
@grondo thanks for the notes and further investigations. Sounds like the likely problem. I think the best solution here is to have the Travis team modify the default `core_pattern` to something that does not require apport. I've done extensive testing of coredump generation recently and have settled on `/tmp/logbt-coredumps/core.%p.%E` has a great default `core_pattern`. It says "put the core file":

  - in a temporary directory readable without root access
  - name the core by pid and program name such that a crash in a binary called `/usr/local/node` with a pid of `34065` would be `/tmp/logbt-coredumps/core.34065.!usr!local!bin!node` (the system escapes `/` as `!` automatically). This is important so that multiple crashed do not overwrite a single coredump.

A script implementing this with extensive tests is at https://github.com/mapbox/logbt. Sadly because `sudo:false` machines do not have a reasonable `core_pattern` I cannot yet use these travis machines, so `sudo: required` is needed until this issue is fixed by the Travis team.

**Comment by Seadragon91** [comment]
I used this two stackoverflow answers to get backtraces if a crash occurs or not:
http://stackoverflow.com/questions/8657648/how-to-have-gdb-exit-if-program-succeeds-break-if-program-crashes/13026450#13026450

With this answer the confirmation on quit from gdb, is switched off:
http://stackoverflow.com/questions/4355978/get-rid-of-quit-anyway-question-just-kill-the-process-and-quit

**Comment by kkm000** [comment]
Just wondering, is it so *impossibly hard* to uninstall apport from host machines (which is useless and is also the reason for this bug), or simply stop the apport daemon, that this issue is dragging on for 2 years, and people are discussing workarounds to workarounds on this ticket?

**Comment by topisani** [comment]
Any updates on this?

**Comment by grooverdan** [comment]
TLDR version, fix sysctl `core_pattern` infrastructure to not be apport.

@slate-bot: so impatient, 3 days to respond so say, "not solved, all information already this bug report".

**Comment by beauby** [comment]
So is there still no way to get a coredump in `sudo: false` environments?

**Comment by pnorman** [comment]
> please do feel free to either reopen this issue

Please reopen this issue.

**Comment by kkm000** [comment]
@joshk @BanzaiMan Yes, this is *still* an issue, and has been for 2 years, 1 month and 7 days (except that the too eager @stale bot closed it. Bad bot! Oh oops. This is not reddit.)

I would really appreciate your comments on this. If just for the sake of this not being sad and ridiculous at the same time.

**Comment by beauby** [comment]
It would make sense to rename the issue to "Enable core dump" or something like that, as the discussion moved away from the initial ulimit addon suggestion.

**Comment by pnorman** [comment]
This request is still relevant, can you please re-open.
